### ISSUE TITLE
[Feature]:  ROCm Using shared memory When GPU runs out of memory

### ISSUE THREAD

**Issue by void-Man** [ROCm/ROCm#3681]
### Suggestion Description

When I paint with ComfyUI, if the workflow is more complex, more models need to be loaded, or larger images are generated, there is a **HIP out of memory**, even if there is still enough space in shared memory at that point.
But when I switch to windows and run ComfyUI with zluda, I find that when the GPU runs out of memory, it will use shared memory to continue running instead of throwing a **HIP out of memory**, I would like to know if there is a plan to support the use of shared memory in subsequent releases.

### Operating System

Ubuntu

### GPU

RX 7900xtx

### ROCm Component

ROCm 6.2.0

**Comment by harkgill-amd** [comment]
Hi @void-Man, are you able to consistently reproduce this issue with a specific model? If so, could you please share the steps to reproduce the `HIP out of memory` error?

**Comment by hartmark** [comment]
I have the same issue, I have 64 GB system ram that sits there unused :(

**Comment by Schwenn2002** [comment]
the same issue:

System:
CPU AMD Ryzen 9950X
RAM 128 GB DDR5
GPU0 AMD Radeon PRO W7900
GPU1 AMD Radeon RX7900XTX
ROCM: 6.3.1
Ubuntu 24.04 LTS (currently patched)

In a multi-GPU setup, only the VRAM of one graphics card is used, the 128GB are not used by ollama (with ROCm) and only an error message: Out of Memory is given.

**Comment by tcgu-amd** [comment]
Hi @void-Man, can you try running your workload with lowvram mode?

Automatically using shared memory when on device memory is insufficient should by no means be a responsibility of ROCm. There is no magic, as whenever shared memory is used, there will be a performance degradation, and we generally do not want to silently introduce a performance degradation at the runtime level. Rather, the user should be informed when device memory is not sufficient via an OOM error, and they could choose what to do in that case. 

In fact, hipMalloc(), which is the API used under the hood to allocate device memory, should always return OOM error if there is not enough memory on device. We actually provide another API, hipMallocManaged() to allow mixed use of both device and host memory, at a non-trivial performance cost. Choosing when to use which API appropriately should be a responsibility of a higher-level framework (e.g. PyTorch and above). 

In your case, choosing when to use shared vs device memory should be managed by ComfyUI itself. This is actually exactly what happens: before an allocation, ComfyUI will calculate the required amount of memory and available memory on the system, and then decide how to use the memory. When you see the OOM error, it is already after ComfyUI has made a decision to allocate on device, and ROCm was just following the orders. An example is here https://github.com/comfyanonymous/ComfyUI/blob/eb2f78b4e09b1970e2fc51fc5d2e062f1a826399/comfy/sd.py#L582, where you can see ComfyUI actually catching the OOM. Another instance is here https://github.com/comfyanonymous/ComfyUI/blob/master/comfy/model_management.py#L618C28-L618C39, where ComfyUI tries to free memory before loading a model if there's not enough memory. 

However, based on your log, OOM actually occurred here https://github.com/comfyanonymous/ComfyUI/blob/d03ae077b4330f58e7caba53ff94e7fd58d0dc7d/comfy/sd.py#L509, which apparently incurs a significant memory cost that is not accounted for by comfyUI. This is a bug on ComfyUI's part, as it should have ensured there will be enough VRAM, but it didn't. 

tl;dr I would say this is rather a ComfyUI issue rather than a ROCm one. As to why different device/platforms may or may not exhibit this issue, it is probably because they way cache was managed and sometimes memory primitives are handled differently, which let ComfyUI get away with miscalculations sometimes. Either fixing on ComfyUI's side or run your workload with lowvram mode, which will allow ComfyUI to use a more reserved memory handling strategy, are better solutions for this issue in my opinion.

**Comment by tcgu-amd** [comment]
@void-Man A bit of additional context, the reason why on windows shared memory is automatically used seems to be a feature of the windows graphics driver as per this post (https://stackoverflow.com/questions/77321986/cudamalloc-not-allocating-to-shared-gpu-memory-on-ampere-micro-architectures-in). As such, Linux support is unlikely, at least not anytime soon.

**Comment by hartmark** [comment]
I have created an issue on ComfyUI, please add more details of you have any

@void-Man

**Comment by void-Man** [comment]
Get it, thank you very much!




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "ROCm/ROCm"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2025年7月25日(星期五) 凌晨2:45
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [ROCm/ROCm] [Feature]:  ROCm Using shared memory When GPU runs out of memory (Issue #3681)



tcgu-amd left a comment (ROCm/ROCm#3681)
 
Hi @void-Man, can you try running your workload with lowvram mode?
 
Automatically using shared memory when on device memory is insufficient should by no means be a responsibility of ROCm. There is no magic, as whenever shared memory is used, there will be a performance degradation, and we generally do not want to silently introduce a performance degradation at the runtime level. Rather, the user should be informed when device memory is not sufficient via an OOM error, and they could choose what to do in that case.
 
In fact, hipMalloc(), which is the API used under the hood to allocate device memory, should always return OOM error if there is not enough memory on device. We actually provide another API, hipMallocManaged() to allow mixed use of both device and host memory, at a non-trivial performance cost. Choosing when to use which API appropriately should be a responsibility of a higher-level framework (e.g. PyTorch and above).
 
In your case, choosing when to use shared vs device memory should be managed ComfyUI itself. This is actually exactly what happens: before an allocation, ComfyUI will calculate the required amount of memory and available memory on the system, and then decide how to use the memory. When you see the OOM error, it is already after ComfyUI has made a decision to allocate on device, and ROCm was just following the orders. An example is here https://github.com/comfyanonymous/ComfyUI/blob/eb2f78b4e09b1970e2fc51fc5d2e062f1a826399/comfy/sd.py#L582, where you can see ComfyUI actually catching the OOM. Another instance is here https://github.com/comfyanonymous/ComfyUI/blob/master/comfy/model_management.py#L618C28-L618C39, where ComfyUI tries to free memory before loading a model if there's not enough memory.
 
However, based on your log, OOM actually occurred here https://github.com/comfyanonymous/ComfyUI/blob/d03ae077b4330f58e7caba53ff94e7fd58d0dc7d/comfy/sd.py#L509, which apparently incurs a significant memory cost that is not accounted for by comfyUI. This is a bug on ComfyUI's part, as it should have ensured there will be enough VRAM, but it didn't.
 
tl;dr I would say this is rather a ComfyUI issue rather than a ROCm one. As to why different device/platforms may or may not exhibit this issue, it is probably because they way cache was managed and sometimes memory primitives are handled differently, which lets ComfyUI get's away with miscalculations sometimes. Either fixing on ComfyUI's side or run your workload with lowvram mode, which will allow ComfyUI to use a more reserved memory handling strategy are better solutions for this issue.

**Comment by tcgu-amd** [comment]
Hi @hartmark, @void-Man, I will be closing this issue for now since there is no action item for ROCm. Please feel free to follow up with more questions though, we will be happy to help answer them. Thanks!
